tests/scripts/simulated-redis-tests.sh

```diff
@@ -0,0 +1,52 @@
+#!/bin/sh
+
+# Usage: $0 /path/to/clusterclient-binary
+
+clientprog=${1:-./clusterclient}
+
+# Sync process just waiting for server to be ready to accept connection.
+perl -we 'use sigtrap "handler", sub{exit}, "CONT"; sleep 5; die "timeout"' &
+syncpid=$!
+
+# Start simulated server
+timeout 5s ./simulated-redis.pl -p 30001 -d --sigcont $syncpid <<EOF &
+EXPECT ["CLUSTER", "SLOTS"]
```

Comment:
Really nice, easy readable now

One options for a testcase could be to combine the testflow into a single place of edit/read
Having a header, and put it in a variable, I guess it could be sent into simulated-redis.pl and $clientprog by first grepping out the lines using.. a bit of fiddling needed for that though..

`
#!/bin/sh
test_sequence=$(cat <<'EOF'
S: EXPECT ["CLUSTER", "SLOTS"]
S: SEND [[0, 16383, ["127.0.0.1", 30001, "nodeid123"]]]
S: EXPECT RECONNECT
C: SET foo bar
S: EXPECT ["SET", "foo", "bar"]
S: SEND +OK
C: GET foo
S: EXPECT ["GET", "foo"]
S: SEND "bar"
S: EXPECT CLOSE
EOF
)

...

server_sequence=\`echo "$test_sequence" | grep '^S:' | cut -d' ' -f2-\`
./simulated-redis.pl .. << EOF
$server_sequence
EOF

...

client_sequence=\`echo "$test_sequence" | grep '^C:' | cut -d' ' -f2-\`
 "$clientprog" .... << EOF
$client_sequence
EOF
`

Reply:
Yes, it's doable, and then the test case becomes just one text file and the script is the test framework, but
* Where do you put the expected output from the clusterclient program? It is also checked in this script.
* How do you write these files when you have multiple simulated Redis nodes?

I think I will write some test case for ASK redirects and then see if it's worth the fiddling. The more fiddling, the harder it is to see what's happening.

What other test cases would be relevant for this kind of tests?

Reply:
Its interesting to have in normal scenarios aswell, but terminating the connection right in the middle of everything. In a transaction sequence for example.

Reply:
I've added ASK redirection test case. I tried adding a MOVED test too, but I can't get it to work..... Maybe tomorrow. :)